source/exe/admin_response.cc

```diff
@@ -0,0 +1,192 @@
+#include "source/exe/admin_response.h"
+
+#include "envoy/server/admin.h"
+
+#include "source/server/admin/admin_filter.h"
+#include "source/server/admin/utils.h"
+
+namespace Envoy {
+
+AdminResponse::AdminResponse(Server::Instance& server, absl::string_view path,
+                             absl::string_view method, SharedPtrSet response_set)
+    : server_(server), opt_admin_(server.admin()), shared_response_set_(response_set) {
+  request_headers_->setMethod(method);
+  request_headers_->setPath(path);
+}
+
+AdminResponse::~AdminResponse() {
+  cancel();
+  shared_response_set_->detachResponse(this);
```

Comment:
I see. I misunderstood and thought that only `MainCommonBase` can outlive `AdminResponse`. (partially due to those attach/detach methods)

 The relationship here actually is: 
(1 ) `MainCommonBase` can outlive `AdminResponse` so that you need those attach/detach methods.
(2)  `AdminResponse` can outlive `MainCommonBase` as well, so you need shared_ptr

Reply:
I'll go ahead and add those two bullets to the doc comment and resolve this PR comment.